# 循环性能比较

- 普通 for 循环 > forEach > for of > map > for in
- for - forEach - of - map - in

# 为什么 break 和 return 不能跳出 forEach ?

```
arr.forEach((value, index, arr) => {})
---

1
原因
forEach 伪代码如下
arr.forEach = function(callback) {
  for(let i = 0; i < this.length; i++) {
    callback(this[i], i, this)
  }
}
所以在 callback 中 break 和 return 都不能跳出 forEach


2
forEach中是不能使用 break 的，会报错，因为forEach是一个函数
- break ------ 结束整个循环，只能用于 for, while, switch 等循环中
- continue --- 结束本次循环，只能用于 for, while, switch 等循环中
- return ----- 结束函数
2.1
arr.forEach((value, index) => {
  if (value > 3) break; // 报错 Jump target cannot cross function boundary
});
2.2
const arr = [1, 2, 3, 4, 5];
arr.forEach((value, index) => {
  console.log(value); // 12345
  if (value > 3) return;
});


3
如果跳出 forEach
- 抛错
- throw new Error('')
- throw new TypeError('')
- 一共有6种错误类型 + 1种自定义错误: TypeError, ReferenceError, RangeError, SyntaxError, UriError, EvalError

arr.forEach((value, index) => {
  console.log(value);
  if (value === 3) {
    throw new Error("");
  }
});
```
